### PR TITLE
Fix: poll DHT in background when worker runs up a workflow + dual-stack webserver

### PR DESCRIPTION
**Pull request by zeeshanlakhani**
Completes:

- Closes https://github.com/ipvm-wg/homestar/issues/278. 
- Closes https://github.com/ipvm-wg/homestar/issues/589.

Includes:
  * We no longer block DHT polling of receipts, and, instead, purposefully race execution and DHT finding/polling.
  * This moves us toward the goal of resolving outside-of-workflow awaited (CIDs), but we'll need another background process to re-run a workflow once known-awaits are resolved (separate ticket). The next set of work will also validate that those outside-of-workflow promises exist, i.e. currently running in another workflow or have already run.
  * v4/v6 host address settings for the Json-RPC web server so that the web server can handle dual v4/v6 requests on the same port.
  * Better / ordered error handling around CIDs that fail to resolve.
  * Workflow status field in the DB (+ migration) (SQLite enum, with special patching).
  * Workflow retries field in the DB (in same migration).
  * A new poller implementation that's general purpose.
  * Removes `async_trait` except for wasmtime impls (due to 1.73).

### CODE REVIEW

File: homestar-runtime/src/scheduler.rs [hunk 1]
```diff
@@ -95,7 +100,6 @@ impl<'a> TaskScheduler<'a> {
     /// [Receipts]: crate::Receipt
     /// [Swarm]: crate::network::swarm
     /// [Workflow]: homestar_workflow::Workflow
-    #[allow(unknown_lints, clippy::needless_pass_by_ref_mut)]
     pub(crate) async fn init<F>(
```
Comment (zeeshanlakhani) [inline homestar-runtime/src/scheduler.rs:103]: add some comments to follow along.

File: homestar-runtime/src/workflow.rs [hunk 2]
```diff
@@ -110,6 +112,38 @@ pub(crate) struct Vertex<'a> {
...
+
+impl Promises {
+    /// Create a new [Promises] object from a given pair of
+    /// in-flow and out-flow [Cid]s.
+    pub(crate) fn new(in_flow: Vec<Cid>, out_flow: Vec<Cid>) -> Promises {
+        Promises { in_flow, out_flow }
+    }
+
+    /// Return an iterator over the [Promises] in-flow and out-flow [Cid]s.
+    #[allow(dead_code)]
```
Comment (zeeshanlakhani) [inline homestar-runtime/src/workflow.rs:139]: check on this.

File: homestar-runtime/src/worker.rs [hunk 3]
```diff
@@ -169,7 +168,50 @@ where
         )
         .await
         {
-            Ok(ctx) => self.run_queue(ctx.scheduler, running_tasks).await,
+            Ok(ctx) => {
+                let promises_to_resolve = ctx.scheduler.promises_to_resolve.clone();
```
Comment (zeeshanlakhani) [inline homestar-runtime/src/worker.rs:172]: possibly ARC.

File: homestar-runtime/src/worker/resolver.rs [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+        }
+    }
+}
+
+impl<DB> Poll<DB> for DHTResolver
+where
+    DB: Database,
+{
+    async fn poll(&self, ctx: &Poller<DB>) -> Result<()> {
+        println!("polling DHTResolver: {:?}", self.cids);
```
Comment (zeeshanlakhani) [inline homestar-runtime/src/worker/resolver.rs:107]: remove this.

File: homestar-runtime/tests/network/dht.rs [hunk 5]
```diff
@@ -315,19 +303,30 @@ fn test_libp2p_dht_records_integration() -> Result<()> {
...
+            vec![
+                "committed to database",
+                "dht.resolver",
+                &put_receipt_cid.to_string(),
+            ],
+        );
+
         assert!(retrieved_workflow_info_logged);
+        assert!(retrieved_receipt_info_logged);
+        assert!(committed_receipt);
```
Comment (zeeshanlakhani) [inline homestar-runtime/tests/network/dht.rs:329]: check the db after the fact.

File: homestar-runtime/src/scheduler.rs (hunk 1 above)
Comment (bgins) [reply homestar-runtime/src/scheduler.rs:103]: Thanks for the extra comments 🙏

**Review by bgins — APPROVED** [review]
Looks good! One test was a bit flaky for me, left a note about it.

File: homestar-runtime/src/workflow.rs [hunk 6]
```diff
@@ -78,8 +78,10 @@ impl fmt::Display for Resource {
 /// [Dag]: dagga::Dag
 /// [Task]: homestar_invocation::Task
 #[derive(Debug, Clone)]
+#[allow(dead_code)]
```
[the remark this answers is not in the document]
Comment (zeeshanlakhani) [reply homestar-runtime/src/workflow.rs:81]: will check.

File: homestar-runtime/src/worker/poller.rs [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+        )
+        .await;
+
+        let msg = rx.recv_async().await.unwrap();
+        assert_eq!(msg, 1);
+        assert!(rx.try_recv().is_err())
+    }
+
+    #[tokio::test]
+    async fn polls_at_interval() {
```
[the remark this answers is not in the document]
Comment (zeeshanlakhani) [reply homestar-runtime/src/worker/poller.rs:119]: ok, will gander.

**Pull request merged**